### ISSUE TITLE
Launcher: Support launching file systems under different accounts

### ISSUE THREAD

**Issue by netheril96** [billziss-gh/winfsp#126]
If I run my program under WinFsp.Launcher, will the program itself run with elevated privileges? If not, will the launcher service be able to execute commands on the behalf of the program with elevated privileges?

**Comment by billziss-gh** [comment]
If you run your file system under the WinFsp.Launcher, it will be run under the SYSTEM account. This is one of the most powerful accounts on the system.

**Comment by netheril96** [comment]
Thanks for the info. It is too dangerous for my program run with this level of power.

**Comment by billziss-gh** [comment]
It depends what it does. But I understand your sentiment.

Perhaps the thing to do here is to add functionality to the launcher so that it can launch file systems under more limited accounts.

**Comment by netheril96** [comment]
Yes, launching file systems under limited accounts is what I need. I am thinking about using the launcher as a service is better than a console application in many ways for a filesystem, but I need the privilege to be restricted so that I can limit potential damage done.

**Comment by billziss-gh** [comment]
Understood. I will reopen, change the title as per your request and mark it as enhancement.

**Comment by billziss-gh** [comment]
**NOTE**: Drives created under the SYSTEM account (using `DefineDosDevice`) are globally visible. Drives created by any other account are only visible to that account. This could be a major impediment in delivering this functionality.

**Comment by billziss-gh** [comment]
Commit 42e01a9b27bc1e67e54ddebaae0718bfcf3a2038 adds this functionality.

Here is how this works. File systems that wish to run under the SYSTEM ("LocalSystem") account will work as before. File systems that do not wish to run under the SYSTEM ("LocalSystem") account may specify an alternative by adding a Registry value `RunAs`. This value may take one of 3 values: [LocalSystem](https://msdn.microsoft.com/en-us/library/windows/desktop/ms684190(v=vs.85).aspx), [LocalService](https://msdn.microsoft.com/en-us/library/windows/desktop/ms684188(v=vs.85).aspx) and [NetworkService](https://msdn.microsoft.com/en-us/library/windows/desktop/ms684272(v=vs.85).aspx).

For example:
```
[HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\WinFsp\Services\memfs64]
"RunAs"="NetworkService"
```

There is currently a big caveat with this. Drives created under the LocalService or NetworkService accounts are **not visible** to any user. This is because drives created with `DefineDosDeviceW` are scoped by the authentication ID as discussed earlier. Only the LocalSystem account has the power to create global drives when using `DefineDosDeviceW`.

This needs to be rectified before this issue is closed. I am looking for more elegant solutions, but right now the only solution that comes to mind is to have the WinFsp.Launcher service create these drives on behalf of the file systems (because it already runs in the LocalSystem context). This is a lot of work and quite inelegant.

**Comment by billziss-gh** [comment]
@netheril96 let me know if running under the LocalService or NetworkService accounts would be enough for your purposes. These accounts are severely restricted.

**Comment by netheril96** [comment]
I was thinking of running the service as the current user, just like a Unix daemon could be doing. Is it not possible on Windows?

**Comment by billziss-gh** [comment]
It is of course possible. And it would also solve the mentioned `DefineDosDeviceW` problem. That was what I thought I would do at first.

But there are some caveats with having services running under a user account. I need to make sure I understand all implications before I allow this.

**Comment by billziss-gh** [comment]
@netheril96 commit 62266ce3b357608f4526f56f9329997122798ccc adds what you want. If you supply `"RunAs"="."` in the registry, the launcher service will "impersonate" the user who starts the file system and will run the file system in the "impersonated" account.

This commit is highly experimental for a number of reasons and may not make it into master. In particular I need to convince myself that there are no security implications/gotchas with this scheme. Frankly I am more comfortable running the file system under a well understood account such as "LocalService" or "NetworkService" and believe the security implications can be better understood in that scenario.

BTW, the `DefineDosDeviceW` problem remains even under this scheme (because the file system is run under session 0 and not the user session -- too complicated to explain without going into too much detail). I need to solve this problem as well.

**Comment by netheril96** [comment]
As I don't under the security model of Windows, I cannot comment on it further. I am just here waiting for you to finalize it.

**Comment by mhogomchungu** [comment]
Following this one closely since it'll be very beneficial to SiriKali.

I also do not understand window's security model and cant add much to this discussion.

**Comment by billziss-gh** [comment]
I am close to finalizing this, but have been distracted by some other things.

To give an overall understanding of the security issues here:

- Running under LocalSystem: LocalSystem is the most powerful account on the computer. I agree with @netheril96 that we should have an option to run file systems under a less powerful account.

- Running as LocalService or NetworkService: On Windows the usual alternative is to run services under LocalService and NetworkService. These are restricted accounts that were created specifically for the purpose of running services in a restricted security context. These accounts are not all powerful and our problem may actually be that they may not have enough permissions for what we want them to do in all cases.

    - For example, it is very likely that I could run the SSHFS-Win file system under LocalService without any problems. But it is unlikely that I could run the passthrough sample without problems, because it accesses the underlying file system and needs permissions to files that LocalService may not have.

- Running as the particular user that wants to start a file system: On UNIX this is easily done by daemonizing a particular process, which then runs in the account that it was created in. On Windows we have to use the launcher infrastructure (for reasons already explained), which means that we have to use the concept of "impersonation".

    - The launcher exposes a named pipe that is used by programs like `launchctl` to communicate with the launcher and control file systems. On Windows it is possible for a named pipe server to get the credentials of a named pipe client and use these credentials for authentication purposes. This is indeed done today by the launcher.

    - It is also possible to use those same credentials for "impersonation" purposes if the client allows it. The named pipe server can then impersonate (somewhat equivalent to the concept of "effective user id" on UNIX) the client and do things like start processes under that account. Unfortunately impersonation opens up a new can of security worms, which is why I am not certain that I want this functionality in WinFsp at this point.

**Comment by mhogomchungu** [comment]
SiriKali on windows now uses launcher to monitor for mount activities and automatically updates its GUI when securefs volumes are unlocked/locked using the launcher infrastructure. Of course securefs is currently not usable this way but it helps in my tests. More progress will follow when the launcher fully supports securefs.

![screenshot 8](https://user-images.githubusercontent.com/3496115/34081155-26ad2c06-e35a-11e7-9923-1071e1af7b3c.png)

**Comment by billziss-gh** [comment]
@mhogomchungu this is fantastic!

I have had a rather busy week and have been unable to finish the discussed functionality, but I am hoping to do so this week.

**Comment by billziss-gh** [comment]
Commit ff155694ce2ac8d3d4c235a5fb96438671710e00 allows services running under an account other than `LocalSystem` to `DefineDosDevice` a global drive.

There has been quite a bit of work around this issue:

1. New registry setting `RunAs`. This can be set to `LocalService` or `NetworkService` to run a file system under one of these (restricted) accounts. (The default is `LocalSystem`, which is the most powerful account in the system.)

2. Experimental `RunAs` value (`.`). There is also the ability to "impersonate" the user account that launches the file system. This might not make it into master, unless we really examine all related security issues. @netheril96 if you really need this (rather than LocalService/NetworkService) please explain your need a bit more.

3. New `DefineDosDevice` functionality. Drives on Windows are local to the user that creates them, unless you are LocalSystem. This is detrimental in our case, because we want services that run under an account other than LocalSystem to create global drives. So I had to build new launcher functionality to accommodate this need.

I will be merging (1) and (3) into master in the near future.

**Comment by billziss-gh** [comment]
I have merged (1) and (3) into master.

**Comment by netheril96** [comment]
My security model follows the default settings of FUSE, that is, the process runs as a current user, and the mounting point is only accessible to the current user.

You have repeatedly mentioned the security issues of impersonating current user, but I don't understand. Could you elaborate on what exactly these security issues might be?

**Comment by mhogomchungu** [comment]
@netheril96, why do you not want securefs to run under "LocalSystem"? You claim to not understand window's security model but seem adamant you do not want securefs to run in this mode.

From the user perspective, the impersonation feature will allow the drive to be accessible only to the impersonated user. Will this drive show up on windows's explorer?

**Comment by netheril96** [comment]
@mhogomchungu I ventured a guess that `securefs` running as `LocalSystem` won't be able to access any files only readable to `SomeUserName`.

**Comment by billziss-gh** [comment]
@netheril96 actually running under LocalSystem your files will more than likely be accessible. Think of LocalSystem as the closest to "root" on Windows.

I thought that your argument was that you did not need to deal with the security implications of running as "root". (Your file system becomes more surface area for attacks to get "root" on Windows.)

In any case I am planning to do a complete security writeup of the Launcher functionality soon. I can then present my reasons for avoiding "impersonation" of users.

**Comment by netheril96** [comment]
@billziss-gh  I had a misunderstanding about what `LocalSystem` can do. In any case, I do want to avoid the security implications of running as "root" and frankly, any privilege greater than the "current user".

**Comment by billziss-gh** [comment]
In this writeup I present the launcher security model. I also discuss potential issues with user "impersonation".

## Overview

The WinFsp launcher is a Windows service named WinFsp.Launcher that runs under the LocalSystem account. Its primary purpose is to launch and manage file system instances. In the latest builds it can also be used to call `DefineDosDevice` from the LocalSystem context, thus creating a global drive.

The launcher creates a "well-known" named pipe upon startup with the name `\\\\.\\pipe\\WinFsp.{14E7137D-22B4-437A-B0C1-D21D1BDF3767}`. All communication with the launcher happens through this named pipe using simple commands like `start`, `stop`, `info`, `list`, etc. The `launchctl` utility can also be used to issue these commands to the launcher from the command line.

## Security

The primary concern with using the launcher named pipe on Windows is that of mutual authentication. The launcher must authenticate the client (e.g. `launchctl`), but the client must ALSO authenticate the launcher.

### Client Authentication

The launcher needs to authenticate the client for obvious reasons: for example, it must be able to check whether a particular user can start a particular file system (controllable through the `Security` registry setting). Windows named pipes support a concept called "impersonation", which allows a pipe server to momentarily assume the identity of the pipe client; the pipe server can use this time to identify the client. This is how this is done in the launcher:

```C
        ClientToken = 0;
        if (!ImpersonateNamedPipeClient(SvcPipe) ||
            !OpenThreadToken(GetCurrentThread(), TOKEN_QUERY, FALSE, &ClientToken) ||
            !RevertToSelf())
        {
            // handle error...
        }
```

This code impersonates the client (`ImpersonateNamedPipeClient`), gets its token (`OpenThreadToken`) and then reverts to its original identity (`RevertToSelf`). The `ClientToken` encapsulates the identity of the pipe client.

Every file system that can be controlled by the launcher must have a set of registry entries under the `HKEY_LOCAL_MACHINE\Software\WinFsp\Services` registry key. (This part of the registry is writable only by an Administrator by default.) One of these entries is `Security` and it controls who can start, stop or get information about a file system. The launcher uses the information in `Security` and compares it against the information in the `ClientToken` to determine whether to allow the requested operation to proceed.

### Server Authentication

While it is obvious that the server must be able to authenticate the client, it may not be as obvious that the client must also be able to authenticate the server.

Recall that the launcher pipe has a well-known name, which means that it can be created by any process in the system! This of course means that an unsuspecting client may connect to what it thinks is the launcher pipe, only to be communicating with a rogue process.

This can have complications in some settings. At the lower (?) threat level a rogue process may be able to observe a user's file system secret (see launcher command `startWithSecret`). At a higher threat level a rogue process may be able to fully impersonate the client that is connecting to it: recall that a pipe server can use `ImpersonateNamedPipeClient` to "become" the client momentarily. If the client does not take precautions, the rogue process may now be able to do things like launch a new process under the client's identity.

The primary defense against this problem is this: the client checks that the launcher pipe is owned by LocalSystem immediately after connecting to it. If it is not owned by LocalSystem it immediately stops trying to access that pipe. This is facilitated by the WinFsp (undocumented) library function `FspCallNamedPipeSecurely`.

There are some additional defenses:

- When the launcher creates the named pipe using `CreateNamedPipe` it uses the `FILE_FLAG_FIRST_PIPE_INSTANCE` and allows only a single instance of the pipe. This ensures that if the launcher is running it is the sole server of the named pipe.

- IMPORTANTLY for our discussion. The client explicitly disallows "impersonation". (The client does this by passing the `SECURITY_SQOS_PRESENT | SECURITY_IDENTIFICATION` flags when opening the pipe in `FspCallNamedPipeSecurely`.) This restricts what the pipe server can do during `ImpersonateNamedPipeClient`. For example, it can use the `ClientToken` to authenticate the user and do access checks, but it cannot use that token to create a new process.

### Allowing Impersonation

In order to achieve the requested functionality the client must allow "full impersonation", where the pipe server can start processes on the client's behalf. If the authentication scheme in `FspCallNamedPipeSecurely` is flawed, there is now the potential for a rogue process to fully impersonate any process/user that attempts to connect to the launcher. This is why I am resisting to allow this.

If the launcher security model is reviewed by people with good understanding of Windows security and "full impersonation" is deemed safe with reasonable confidence, then I will be happy to add it.

----

### Aside: Windows Impersonation Levels

A quick aside on Windows impersonation levels. Windows has 4 impersonation levels:

- Anonymous: does not allow any impersonation. The server has no knowledge of the client.
- Identification: allows identification but not impersonation. The server can identify the client (e.g. for authentication purposes), but it cannot do tasks like `CreateProcess` on its behalf. This is what WinFsp uses today, as discussed above.
- Impersonation: allows full impersonation. The server can fully "become" the client and launch processes under the client's credentials.
- Delegation: impersonation on a remote computer. As scary as it sounds.

**Comment by netheril96** [comment]
If my understanding is not wrong, the client process and the launcher service are separate entities, and the former always runs in the context of current user, while the latter by default of `LocalSystem`. In that case, I could do without impersonation, if the launcher could deny any user other than client user access to the filesystem, just like FUSE.

**Comment by mhogomchungu** [comment]
I am doing the impersonation thing in my zuluCrypt project[1] and hence i understand the problem you are trying to solve. All those "uid" and commands with "uid" in their names are there to keep track of under what user account the executable is running on and it is running to serve what user.

Communication between unprivileged process and a privileged process is done via a known path of a unix local socket and the server authenticates the client by checking a cookie it knows against a cookie a client submitted[2]. Too bad i can not assist you with your impersonation because i do not know windows programming.

```
If the authentication scheme in FspCallNamedPipeSecurely is flawed, there is now 
the potential for a rogue process to fully impersonate any process/user 
that attempts to connect to the launcher. This is why I am resisting to allow this.
```
One way to prevent this is to require launcher's clients to:
1. Register with the launcher in sections of the registry where only privileged users have access to. I believe this is already required.
2. Have client's executables in known and supported paths only. For example,you may require them to be in a subfolder of winfsp installation folder or in any other system path where normal users have no write access too.

[1] https://github.com/mhogomchungu/zuluCrypt/blob/01aeb4f23a9e5d95515a02dbe47b1e8031ef0dd3/zuluCrypt-cli/bin/main.c#L353

[2] https://github.com/mhogomchungu/zuluCrypt/blob/01aeb4f23a9e5d95515a02dbe47b1e8031ef0dd3/zuluPolkit/zulupolkit.cpp#L227

**Comment by mhogomchungu** [comment]
```
If my understanding is not wrong, the client process and the launcher service 
are separate entities, and the former always runs in the context of current user, 
while the latter by default of LocalSystem
```
I believe windows works the same way linux does in this regard where when a process starts a sub process, the subprocess by default runs with the same security context/user account as the parent and additional steps must be taken to make the sub process run with different security context/user account.

It is generally a bad idea for a privileged process to manage unprivileged processes and other resources managed by unprivileged users and billziss-gh is justified in having concerns.

FUSE already does this impersonation and this is why its not noticeable unless you look for it like noticing that "fusermount" is suid-root.

**Comment by billziss-gh** [comment]
Commit 3119922708a6004d28ef58a464cd120e424b69fa finally adds the ability to "RunAs" the user who maps the network drive (instead of simply LocalSystem, LocalService or NetworkService). Although I have gone back and forth on this, I expect this change to make into v1.4 unless a major problem is discovered.

**Comment by billziss-gh** [comment]
It should be noted that file systems that use `RunAs="."` are somewhat strange beasts:

- They run in Session 0 (i.e. the Windows Services session).
    - Kernel objects (e.g. events, semaphores, etc.) will be placed in `\BaseNamedObjects`.
    - Window objects (window stations, desktops, etc.) will be placed in `\Windows`.
    - This means that such file systems cannot interact directly with applications in the "interactive" sessions (Session 1, etc.), because interactive session objects go to `\Sessions\{N}\BaseNamedObjects` and `\Sessions\{N}\Windows`.
- Their `DosDevices` are placed in the logon session of the user that launches the file system:
    - This allows drives created by the file system to be visible by the user's Explorer **without creating global drives**.

For example, consider a file system that creates a kernel event `MyEvent` (using `CreateEventW`), is configured with `RunAs="."`, and is launched from the user's (**non-elevated!**) command prompt with `net use X: \\myfs\share`.

- `MyEvent` will be placed in `\BaseNamedObjects` (and not in `\Sessions\1\BaseNamedObjects`).
- The `X:` drive will be placed in `\Sessions\0\DosDevices\{LUID}`, where `LUID` is the logon session of the user who run the `net use` command. (Please note that the drive will be visible from Explorer only if the `net use` command is run from a non-elevated prompt. When using an elevated prompt the `LUID` changes and a non-elevated Explorer cannot find the drive.)

----

I include here my understanding of how Windows (latest versions) maps objects across sessions. This assumes some familiarity with the NTOS namespace. The symbol `->` means a symbolic link.

- Session 0: Services
    - "DOS" devices (e.g. drives):
    ```
    \Sessions\0\DosDevices\{LUID}
    ```
    - Kernel objects (e.g. events, etc.):
    ```
    \Sessions\0\BaseNamedObjects -> \BaseNamedObjects
    ```
    - Window objects (e.g. window stations, desktops, windows, etc.)
    ```
    \Sessions\0\Windows -> \Windows
    ```

- Session 1: Interactive Apps
    - "DOS" devices (e.g. drives) also go to **`\Sessions\0\DosDevices`**. This is not a typo! `\Sessions\1\DosDevices` is not used.
    ```
    \Sessions\0\DosDevices\{LUID}
    ```
    - Kernel objects (e.g. events, etc.):
    ```
    \Sessions\1\BaseNamedObjects
    ```
    - Window objects (e.g. window stations, desktops, windows, etc.)
    ```
    \Sessions\1\Windows
    ```

Because of the reuse of `\Sessions\0\DosDevices` you can have a Session 0 and a Session 1 process using the same `LUID` (Logon Session) and therefore the same `DosDevices`.

**Comment by billziss-gh** [comment]
The `RunAs="."` functionality is now merged into the master and release/1.4 branches. The change should appear in the next `v1.4` release. Closing this.

**Comment by billziss-gh** [comment]
The latest release includes the `RunAs="."` functionaliity.

https://github.com/billziss-gh/winfsp/releases/tag/v1.4B4
